<h3 style="text-align:center;">ESCUELA POLITÉCNICA NACIONAL</h3>
<h3 style="text-align:center;">FACULTAD DE INGENIERÍA EN SISTEMAS</h3>
<h3 style="text-align:center;">METODOS NUMERICOS ICCD412 </h3>

*Sebastián Pazmiño*

*TAREA N 1*

*GR1CC*


*Conjunto de Ejercicios 1*

***Resuelva los siguientes ejercicios, tome en cuenta que debe mostrar el desarrollo completo del ejercicio***

***Formulas para error-relativo y error-absoluto***

### $erro_{abs}= | p - p^* |$

### $ error_{rela} = | \frac{p - p^*}{p}| , \ne 0$

***-a.  $p= \pi , p^*= 22/7$***

***-b.  $p= e , p^*= 2.718$*** 

***-c.  $p= \pi , p^*= 3.1416$*** 

***-d.  $p= \sqrt2 , p^*= 1.414$*** 


*En la resolucion de estos ejercicios se hara uso de la libreria math en python esto para la resolucion de operaciones matematicas avanzada como las Raíces cuadradas potencia y logaritmos tambien se las usara para las constantes matematicas y Redondeo y valores absolutos.*


In [2]:
import math
euler = math.e
pi = 3.14159265

# ************* a *************
p = pi
p1 = 22/7

error_abs = abs(p - p1)
erro_rela = abs((p - p1)/p)
print("error absoluto =", error_abs , "error relativo =", erro_rela)

# ************ b *************
p1 = 2.718

error_abs = abs(p - p1)
erro_rela = abs((p - p1)/p)
print("error absoluto =", error_abs , "error relativo =", erro_rela)

# ************ c *************  
p = pi
p1 = 3.1416

error_abs = abs(p - p1)
erro_rela = abs((p - p1)/p)
print("error absoluto =", error_abs , "error relativo =", erro_rela)

# ************ d *************      
p = math.sqrt(2)
p1 = 1.414

error_abs = abs(p - p1)
erro_rela = abs((p - p1)/p)
print("error absoluto =", error_abs , "error relativo =", erro_rela)

error absoluto = 0.001264492857142585 error relativo = 0.0004025005778971965
error absoluto = 0.42359265000000024 error relativo = 0.13483372836386034
error absoluto = 7.349999999739509e-06 error relativo = 2.339577666041302e-06
error absoluto = 0.00021356237309522186 error relativo = 0.00015101140222192286


*2. Calcule los errores absoluto y relativo en las aproximaciones de  $p$ por $p^*$*

- a. $p = e^{10} , p^* = 22000$

- b. $p = 10^{\pi} , p^* = 1400$

- c. $p = 8! , p^* = 39900$

- d. $p! , p^s* = \sqrt{18\pi}(9/e)^{9} $

 

In [3]:
import math

constante_euler = math.e
pi = 3.14159265

# ********** a **********   
valor_real_a = constante_euler ** 10
valor_aproximado_a = 22000

error_absoluto_a = abs(valor_real_a - valor_aproximado_a)
error_relativo_a = abs((valor_real_a - valor_aproximado_a) / valor_real_a)
print("error absoluto =", error_absoluto_a, "error relativo =", error_relativo_a)

# ********* b **********     
valor_real_b = 10 ** pi
valor_aproximado_b = 1400

error_absoluto_b = abs(valor_real_b - valor_aproximado_b)
error_relativo_b = abs((valor_real_b - valor_aproximado_b) / valor_real_b)
print("error absoluto =", error_absoluto_b, "error relativo =", error_relativo_b)

# ******** c **********     
valor_real_c = math.factorial(8)
valor_aproximado_c = 39900

error_absoluto_c = abs(valor_real_c - valor_aproximado_c)
error_relativo_c = abs((valor_real_c - valor_aproximado_c) / valor_real_c)
print("error absoluto =", error_absoluto_c, "error relativo =", error_relativo_c)

# ******** d **********     
valor_real_d = math.factorial(9)
valor_aproximado_d = math.sqrt(18 * pi) * (9 / constante_euler) ** 9

error_absoluto_d = abs(valor_real_d - valor_aproximado_d)
error_relativo_d = abs((valor_real_d - valor_aproximado_d) / valor_real_d)
print("error absoluto =", error_absoluto_d, "error relativo =", error_relativo_d)

error absoluto = 26.465794806703343 error relativo = 0.0012015452253326688
error absoluto = 14.544280084894353 error relativo = 0.010497831057195794
error absoluto = 420 error relativo = 0.010416666666666666
error absoluto = 3343.127363467007 error relativo = 0.009212762796150261


*3. Encuentre el intervalo más largo en el que se debe encontrar $p^*$ para aproximarse a p con error relativo máximo de 10^-4 para cada valor de p*

- Sabemos que el error relativo es 

- $ error_{rela} = | \frac{p - p^*}{p} | \leq 10^{-4}$

- entonces por lo tanto tenemos que 

- $ | p -p^* | \leq 10^{-4} |p| $

- ahora como p > 0 , tenemo que $ p(1 - 10^{-4}) \leq p^* p(1 +10^{-4})$

**- a. $\pi$**

**- b. $e$**

**- c. $\sqrt2$**

**- d. $\sqrt[3]{7}$**







In [4]:
import math

constante_euler = math.e
pi = 3.14159265

# ***************** a ******************
valor_real_a = pi

limite_inferior_a = valor_real_a * (1 - 10**-4)
limite_superior_a = valor_real_a * (1 + 10**-4)

print("Intervalo = [", limite_inferior_a, ",", limite_superior_a, "]")

# ******************* b ******************
valor_real_b = constante_euler

limite_inferior_b = valor_real_b * (1 - 10**-4)
limite_superior_b = valor_real_b * (1 + 10**-4)

print("Intervalo = [", limite_inferior_b, ",", limite_superior_b, "]")

# ************** c  *********************
valor_real_c = math.sqrt(2)

limite_inferior_c = valor_real_c * (1 - 10**-4)
limite_superior_c = valor_real_c * (1 + 10**-4)

print("Intervalo = [", limite_inferior_c, ",", limite_superior_c, "]")

# ************** d  *********************
valor_real_d = 7 ** (1/3)

limite_inferior_d = valor_real_d * (1 - 10**-4)
limite_superior_d = valor_real_d * (1 + 10**-4)

print("Intervalo = [", limite_inferior_d, ",", limite_superior_d, "]")

Intervalo = [ 3.1412784907350004 , 3.141906809265 ]
Intervalo = [ 2.718010000276199 , 2.718553656641891 ]
Intervalo = [ 1.4140721410168577 , 1.4143549837293325 ]
Intervalo = [ 1.9127398896541117 , 1.9131224758906662 ]


*4. Use la aritmética de redondeo de tres dígitos para realizar lo siguiente. Calcule los errores absoluto y relativo
con el valor exacto determinado para por lo menos cinco dígitos*

**-a. $ \frac{\frac{13}{14}  - \frac{5}{7}}{2e-5.4} $**

**-b. $ -10\pi +6e - \frac{3}{61} $**

**-c. $ (\frac{2}{9}) \cdot (\frac{9}{11}) $**

**-d. $ \frac{|sqrt{13} + \sqrt{11}}{\sqrt{13} - \sqrt{11}} $**



In [5]:
import math
euler = math.e
pi = 3.14159265

# ************* a ************
a = (13/14 - 5/7)/(2*euler-5.4)

p = round(a,3)
p1= round(a,5)
print(p , p1 )

errorabs = abs(p - p1)
errorela = abs((p - p1)/p)
print("error absoluto =",  round(errorabs,5) , "error relativo =", round(errorela,8))

# ************* b ************
a = -10*pi + 6*euler - 3/61

p = round(a,3)
p1= round(a,5)
print(p , p1 )

errorabs = abs(p - p1)
errorela = abs((p - p1)/p)
print("error absoluto =",  round(errorabs,5) , "error relativo =", round(errorela,8))


# ************* c ************
a = (2/9) * (9/11)

p = round(a,3)
p1= round(a,5)
print(p , p1 )

errorabs = abs(p - p1)
errorela = abs((p - p1)/p)
print("error absoluto =",  round(errorabs,5) , "error relativo =", round(errorela,4))

# ************* d ************
a = (math.sqrt(13)+math.sqrt(11))/(math.sqrt(13)-math.sqrt(11))

p = round(a,3)
p1= round(a,5)
print(p , p1 )

errorabs = abs(p - p1)
errorela = abs((p - p1)/p)
print("error absoluto =",  round(errorabs,5) , "error relativo =", round(errorela,8))

5.861 5.86062
error absoluto = 0.00038 error relativo = 6.484e-05
-15.155 -15.15542
error absoluto = 0.00042 error relativo = 2.771e-05
0.182 0.18182
error absoluto = 0.00018 error relativo = 0.001
23.958 23.95826
error absoluto = 0.00026 error relativo = 1.085e-05


*5. Los primeros tres términos diferentes a cero de la serie de Maclaurin para la función arcotangente son: $ x - (1/3)x^{3} + (1/5)x^{5} $ Calcule los errores absoluto y relativo en las siguientes aproximaciones de 𝜋 mediante el polinomio en lugar del arcotangente:*

**-a. $ 4 [ arctan (\frac{1}{2}) + arctan (\frac{1}{3})]$**

**-b. $ 16  arctan (\frac{1}{5}) - 4 arctan (\frac{1}{239})$**

In [5]:
import math

pi = math.pi

# *************** a ****************
print("a)")
valor_x1_a = 1/2
termino_arctan_x1_a = valor_x1_a - (1/3)*valor_x1_a**3 + (1/5)*valor_x1_a**5

valor_x2_a = 1/3
termino_arctan_x2_a = valor_x2_a - (1/3)*valor_x2_a**3 + (1/5)*valor_x2_a**5

valor_real_pi = pi
aproximacion_pi_a = 4 * (termino_arctan_x1_a + termino_arctan_x2_a)

print("valor real:", valor_real_pi, "valor aproximado:", aproximacion_pi_a)

error_absoluto_a = abs(valor_real_pi - aproximacion_pi_a)
error_relativo_a = abs((valor_real_pi - aproximacion_pi_a) / valor_real_pi)
print("error absoluto =", error_absoluto_a, "error relativo =", error_relativo_a)

# *************** b ****************
print("b)")
valor_x1_b = 1/5
termino_arctan_x1_b = valor_x1_b - (1/3)*valor_x1_b**3 + (1/5)*valor_x1_b**5

valor_x2_b = 1/239
termino_arctan_x2_b = valor_x2_b - (1/3)*valor_x2_b**3 + (1/5)*valor_x2_b**5

valor_real_pi = pi
aproximacion_pi_b = (16 * termino_arctan_x1_b - 4 * termino_arctan_x2_b)

print("valor real:", valor_real_pi, "valor aproximado:", aproximacion_pi_b)

error_absoluto_b = abs(valor_real_pi - aproximacion_pi_b)
error_relativo_b = abs((valor_real_pi - aproximacion_pi_b) / valor_real_pi)
print("error absoluto =", error_absoluto_b, "error relativo =", error_relativo_b)

a)
valor real: 3.141592653589793 valor aproximado: 3.1455761316872426
error absoluto = 0.003983478097449478 error relativo = 0.0012679804598147663
b)
valor real: 3.141592653589793 valor aproximado: 3.1416210293250346
error absoluto = 2.837573524150372e-05 error relativo = 9.032277055104427e-06


*6. El número $e$ se puede definir por medio de $ e = \sum_{n=0}^{\infty} (1/n!) $ , donde $n! = n(n-1) ... 2 \cdot 1$ para $n \ne 0$ y $ 0! = 1$. Calcule los errores absoluto y relativo en la siguiente aproximación de $e$:*

**-a. $\sum_{n=0}^{5} (1/n!)$**

**b. $\sum_{n=0}^{10} (1/n!)$**

In [7]:
import math
euler = math.e
# ************* a ************
print("a)")
p=euler
p1= sum(1 / math.factorial(n) for n in range(6))

print("valor real:",p ,"valor aproximado:", p1 )

errorabs = abs(p - p1)
errorela = abs((p - p1)/p)
print("error absoluto =",  errorabs , "error relativo =", errorela)

# ************** b ************
print("b)")
p=euler
p1= sum(1 / math.factorial(n) for n in range(11))

print("valor real:",p ,"valor aproximado:", p1 )

errorabs = abs(p - p1)
errorela = abs((p - p1)/p)
print("error absoluto =",  errorabs , "error relativo =", errorela)

a)
valor real: 2.718281828459045 valor aproximado: 2.716666666666667
error absoluto = 0.0016151617923783057 error relativo = 0.0005941848175815963
b)
valor real: 2.718281828459045 valor aproximado: 2.7182818011463845
error absoluto = 2.7312660577649694e-08 error relativo = 1.0047766310211053e-08


*7. Suponga que dos puntos $(x_0 , y_0) (x_1 , y_1)$  se encuentran en línea recta con $y_1 \ne y_0$. Existen dos fórmulas para encontrar la intersección $𝑥$ de la línea:*

$ x = \frac{x_0y_1 - x_1y_0}{y_1 - y_0} $ y $ x = x_0 - \frac{(x_1 - x_0)y_0}{y_1 - y_0}$

*a. Use los datos $(x_0 , y_0) = (1.31 , 3.24)$ y $(x_1 , y_1) = (1.93 , 5.76)$  y la aritmética de redondeo de tres dígitos para calcular la intersección con $x$ de ambas maneras. ¿Cuál método es mejor y por qué?*

### *Formula 1.* 

$ x = \frac{x_0y_1 - x_1y_0}{y_1 - y_0} $

$ x = \frac{(1.31) \cdot (5.76) - (1.93)\cdot (3.24)}{5.76 - 3.24} $

$ x = \frac{7.546 - 6.253}{2.52} $

$ x = \frac{1.293}{2.52} = 0.513$ 

### *Formula 2.*

$ x = x_0 - \frac{(x_1 - x_0)y_0}{y_1 - y_0}$

$ x = 1.31 - \frac{(1.93 - 1.31)3.24}{5.76 - 3.24}$

$ x = 1.31 - \frac{(0.62)3.24}{2.52}$
    
$ x = 1.31 - \frac{2.009}{2.52}$

$ x = 1.31 - 0.797 = 0.513$

In [8]:
x0= 1.31
y0= 3.24

x1= 1.93 
y1= 5.76

# primera fórmula
a = (x0*y1 - x1*y0)/(y1-y0)
# Segunda fórmula
b=  x0 - ((x1-x0)*y0)/(y1-y0)


print("Primera Formula", round(a,3))
print("Segunda Formula", round(b,3))


Primera Formula 0.513
Segunda Formula 0.513


*La segunda fórmula opera de manera más eficiente y con mayor estabilidad numérica. También se observa que reduce la propagación de errores, como aquellos derivados del redondeo durante los cálculos. Esta aproximación evita operaciones potencialmente problemáticas como ciertas multiplicaciones que pueden amplificar los errores. Además, se destaca por su estructuración de operaciones diseñada para preservar la precisión en los cálculos.*

*Es importante señalar que ambas fórmulas son equivalentes en teoría y producen el mismo resultado exacto. Sin embargo, podrían generar diferencias numéricas cuando se trabajan con un mayor número de dígitos o en implementaciones computacionales específicas, debido a la distinta forma en que manejan los errores de precisión.*

<p><center>Insertar ecuaciones</center></p>

| **Nombre**               | **Comando**                         | **Salida**                      |
|--------------------------|-------------------------------------|---------------------------------|
| Número simple            | `$123$`                             | $123$                           |
| Decimal                  | `$3.14$`                            | $3.14$                          |
| Fracción                 | `$\frac{a}{b}$`                     | $\frac{a}{b}$                   |
| Potencia (exponente)     | `$x^2$`                             | $x^2$                           |
| Raíz cuadrada            | `$\sqrt{9}$`                        | $\sqrt{9}$                      |
| Raíz n-ésima             | `$\sqrt[3]{8}$`                     | $\sqrt[3]{8}$                   |
| Notación científica      | `$5 \times 10^6$`                   | $5 \times 10^6$                 |
| Multiplicación           | `$a \times b$`                      | $a \times b$                    |
| División (símbolo ÷)     | `$a \div b$`                        | $a \div b$                      |
| Igual                    | `$a = b$`                           | $a = b$                         |
| No igual                 | `$a \neq b$`                        | $a \neq b$                      |
| Menor que                | `$a < b$`                           | $a < b$                         |
| Mayor que                | `$a > b$`                           | $a > b$                         |
| Menor o igual que        | `$a \leq b$`                        | $a \leq b$                      |
| Mayor o igual que        | `$a \geq b$`                        | $a \geq b$                      |
| Aproximadamente igual    | `$a \approx b$`                     | $a \approx b$                   |
| Más-menos (±)            | `$a \pm b$`                         | $a \pm b$                       |
| Porcentaje               | `$50\%$`                            | $50\%$                          |
| Proporcionalidad         | `$a \propto b$`                     | $a \propto b$                   |
| Pertenencia (∈)          | `$a \in B$`                         | $a \in B$                       |
| No pertenencia (∉)       | `$a \notin B$`                      | $a \notin B$                    |
| Suma                     | `$\sum_{i=1}^{n} i$`                | $\sum_{i=1}^{n} i$              |
| Producto (∏)             | `$\prod_{i=1}^{n} i$`               | $\prod_{i=1}^{n} i$             |
| Límite                   | `$\lim_{x \to 0} f(x)$`             | $\lim_{x \to 0} f(x)$           |
| Derivada simple          | `$\frac{d}{dx}f(x)$`                | $\frac{d}{dx}f(x)$              |
| Derivada parcial         | `$\frac{\partial}{\partial x}f(x,y)$`| $\frac{\partial}{\partial x}f(x,y)$ |
| Integral                 | `$\int_a^b f(x)\,dx$`               | $\int_a^b f(x)\,dx$             |
| Paréntesis automáticos   | `$\left( \frac{a}{b} \right)$`      | $\left( \frac{a}{b} \right)$    |
| Corchetes automáticos    | `$\left[ \frac{a}{b} \right]$`      | $\left[ \frac{a}{b} \right]$    |
| Llaves automáticas       | `$\left\{ \frac{a}{b} \right\}$`    | $\left\{ \frac{a}{b} \right\}$  |
| Vector                   | `$\vec{v}$`                         | $\vec{v}$                       |
| Matriz simple            | `$\begin{matrix}a & b\\c & d\end{matrix}$` | $\begin{matrix}a & b\\c & d\end{matrix} $|


**link del repositorio de git-hub**

https://github.com/Sebas-p2304/Metodos-Numericos